### DISCUSSION TITLE
请问有没有子查询分页更好的实现方法

### DISCUSSION THREAD

**Discussion by infFox** [dotnetcore/FreeSql#1174] [Q&A]
想要达到的效果类似于：
```sql
SELECT id, name FROM product WHERE id >= (SELECT id FROM product LIMIT 100, 1) LIMIT 0, 20
```

目前实现的方法是：
```csharp
var result = _storeDB.Select<DbProduct>()
         .Where(s => s.Deletedtime == 0)
         .Where("id >= @Id", new { Id = _storeDB.Select<DbProduct>().OrderBy(a => a.Id).Offset(100).Limit(1).ToList<int>(a => a.Id)[0] })
         .OrderBy(a => a.Id)
         .Limit(20)
         .ToList<DbProduct>("id,name");
```

通过 `UseMonitorCommand` 看，好像最终分成 2 个不同的 SQL 语句执行了

请问有没有更好的实现方法呢，能实现类似于最上边那种内联进 WHERE 的效果。谢谢。

**Comment by 2881099** [comment]
.Where(a=> a.Id >= 

    fsql.Select\<DbProduct>().Offset(200).OrderBy(b=> b.Id).First(b => b.Id)

)

**Reply by infFox** [reply]
> .Where(a=> a.Id >=
> 
> ```
> fsql.Select\<DbProduct>().Offset(200).OrderBy(b=> b.Id).First(b => b.Id)
> ```
> 
> )

谢谢！试了一下成功了

之前尝试的方法是
```csharp
.Where(s => s.Id >= _storeDB.Select<DbProduct>().OrderBy(a => a.Id).Offset(200).Limit(1).ToList<int>(a => a.Id)[0])
```

结果运行时提示

> Exception: 未实现函数表达式 value(FreeSqlTest.Controllers.TestController)._storeDB.Select().OrderBy(a => a.Id).Offset(200).Limit(1).ToList(a => a.Id).get_Item(0) 解析
>     
> FreeSql.Internal.CommonExpression.ExpressionLambdaToSql(Expression exp, ExpTSC tsc)
> FreeSql.Internal.CommonExpression.ExpressionBinary(string oper, Expression leftExp, Expression rightExp, ExpTSC tsc)
> FreeSql.Internal.CommonExpression.ExpressionLambdaToSql(Expression exp, ExpTSC tsc)
> FreeSql.Internal.CommonExpression.ExpressionWhereLambda(List<SelectTableInfo> _tables, Func<Type, string, string> _tableRule, Expression exp, BaseDiyMemberExpression diymemexp, List<Item> whereGlobalFilter, List<DbParameter> dbParams)
> FreeSql.Internal.CommonProvider.Select0Provider<TSelect, T1>.InternalWhere(Expression exp)
> FreeSql.Internal.CommonProvider.Select1Provider<T1>.WhereIf(bool condition, Expression<Func<T1, bool>> exp)
> FreeSql.Internal.CommonProvider.Select1Provider<T1>.Where(Expression<Func<T1, bool>> exp)
> FreeSqlTest.Controllers.TestController.Get() in TestController.cs  `var result = _storeDB.Select<DbProduct>()`
> lambda_method5(Closure , object )
> Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor+AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, object controller, object[] arguments)
> ...

看来应该使用 .First()
